### 1.Variable and memory references

There are two components in a register:
1) Storage location
2) Memory address(which is continuous)

- Memory address is shown in hexadecimal.

- Variable is known as name in python

In [1]:
a

NameError: name 'a' is not defined

In [2]:
a = 4
id(a)

1873031227792

In [4]:
hex(1873031227792)

'0x1b4195c6990'

In [5]:
id(4)

1873031227792

- Call by object reference is when name is pointing towards to the value saved in memory address.

### 2.Aliasing

In [8]:
a = 5
b = a
id(a)

1873031227824

In [7]:
id(b)

1873031227824

In [9]:
c = b
id(c)

1873031227824

In [10]:
del a

In [11]:
id(a)

NameError: name 'a' is not defined

In [12]:
print(b)

5


In [13]:
del b

In [14]:
print(c)

5


In [16]:
a = 5
b = a
a = 6
print(b)

5


### 3.Reference counting


- To count the number of variables pointing towards the memory,Reference counting is used.

In [17]:
import sys

In [18]:
a = 'corona'
b = a
c = b
id(a)

1873112442992

In [19]:
id(b)

1873112442992

In [20]:
id(c)

1873112442992

In [21]:
sys.getrefcount(a)

4

- The total number of reference count is total variable pointing towards the memory + 1(because while checking it also points towards the memory address)

In [26]:
a = 'vartika'
b = a
c = b
sys.getrefcount(a)

5

### 4.Garbage collection

- Python doesn't provide the feature of deleting anything from the memory address.It checks whether any variable is pointing towards the memory.The inbuilt program of python called Garbage Collector,checks periodically, whether any variable is pointing towards the memory.If unused,it free ups  the space for further computation.This process is known as Garbage collection.

### 5.Weird behaviours of Python

##### WB 1.

In [27]:
a = 2
b = a
c = b
sys.getrefcount(a)

1677

- There are 1677 extra variables pointing towards the memory address.

In [31]:
a = 5354
b = a  
c = b
d = c
sys.getrefcount(a)

6

In [32]:
d = c
sys.getrefcount(a)

5

In [33]:
a = 4
b = 4
id(a)

1873031227792

In [34]:
id(b)

1873031227792

##### WB 2.

a = 257
b = 257
id(a)
id(b)

In [36]:
id(a)

1873112358128

In [37]:
id(b)

1873112357072

In [40]:
a = -6
id(a)

1873112358192

In [41]:
b = -6
id(b)

1873112356976

In [43]:
# <-5 to 256< (different ids are generated)
# this is known as software optimisation.The numbers in this range is used often

##### WB3.

In [45]:
a = 'haldia'
id(a)

1873113240816

In [46]:
b = 'haldia'
id(b)

1873113240816

In [47]:
a = 'haldia inst tech'
id(a)

1873113218064

In [48]:
b = 'haldia inst tech'
id(b)

1873113256944

In [49]:
a = 'haldia_inst_tech'
id(a)

1873113256384

In [50]:
b = 'haldia_inst_tech'
id(b)

1873113256384

In [51]:
# string should be valid identifier(shouldn't start with number and shouldn't have space)

In [52]:
# valid identifier same identities


### 6.Mutability

In [53]:
L = [1,2,3]
id(L)

1873113241216

In [54]:
id(1)

1873031227696

In [55]:
id(L[0])

1873031227696

In [56]:
L[2] = 1
print(L)

[1, 2, 1]


In [57]:
id(L[2])

1873031227696

- Mutability refers to the ability to change or edit data in it's memory location

- Mutability depends on the datatype


In [58]:
a = 'Hello'
id(a)

1873112384880

In [2]:
a = 'Hello'
a =a +'World'
print(a)

HelloWorld


In [3]:
id(a)

1905531528240

In [4]:
T = (1,2,3)
id(T)

1905526154240

In [5]:
T = T + (5,6)
print(T)

(1, 2, 3, 5, 6)


In [6]:
id(T)

1905530537360

In [7]:
L = [1,2,3]
id(L)

1905531579136

In [8]:
L.append(4)
print(L)

[1, 2, 3, 4]


In [9]:
id(L)

1905531579136

In [10]:
# Side effects of mutability

In [18]:
L = [1,2,3]
L1 = L
print(L)
print(L1)

[1, 2, 3]
[1, 2, 3]


In [19]:
id(L)

1905531555328

In [20]:
id(L1)

1905531555328

In [21]:
L1.append(4)
print(L1)

[1, 2, 3, 4]


In [22]:
id(L1)

1905531555328

In [23]:
L1 = L[:]
id(L1)

1905531561472

In [24]:
id(L)

1905531555328

In [25]:
# This is known as cloning

In [26]:
a = (1,2,3),[4,5]
print(a)

((1, 2, 3), [4, 5])


In [28]:
a[-1][-1] = 500
print(a)

((1, 2, 3), [4, 500])


In [29]:
a = [1,2,3,(4,5)]
a[-1][-1] = 500

TypeError: 'tuple' object does not support item assignment

In [30]:
a = [1,2]
b = [3,4]
c = (a,b)
id(a)

1905531579136

In [31]:
id(b)

1905532372864

In [32]:
id(c)

1905530391808

In [35]:
c[0][0] = 100
print(c)

([100, 2], [3, 4])


In [36]:
id(a)

1905531579136

In [37]:
id(c)

1905530391808

In [40]:
L = [1,2,3]
id(L)

1905532384384

In [2]:
L = [1,2,3]
L = L + [4,5]
print(L)

[1, 2, 3, 4, 5]
